com.unity.ml-agents/Runtime/Agent.cs

```diff
@@ -519,6 +536,14 @@ protected virtual void OnDisable()
             m_Initialized = false;
         }
 
+        void OnDestroy()
```

Comment:
I think we want to put this in OnDisable. I think the user would easily make a custom OnDestroy implementation for their Agents (overriding this)

Reply:
I thought about that before but that means the user has to link the agent with the same TeamManager again every time the agent is disabled and re-enabled. So I thought putting it in destroy might be lesser evil. For existing codes that are not using TeamManager, overriding it is fine.
Also if you want to do something to each agent in OnTeamEpisodeBegin() for next episode, the unregistered agent might not be there.

But that's a good point. Is adding documentation telling user to call base.Destroy() enough or you think we should do it in OnDisable?

Reply:
I don't think it is wise to expect the user will consistently call base.Destroy(. I think we should keep all of our logic in OnDisable (which is also called when the Agent is Destroyed).
I really think we should not introduce any distinction between disabling and destroying the Agent. both should be doing the same thing (as it is the case today). We should not introduce a distinction now.

Reply:
> expect the user will consistently call base.Destroy

We already expect that with OnEnable() and OnDisable() don't we?

It's a bit questionable to me about whether we want to introduce the distinction. We already introduce some distinctions to the workflows and we decoupled disabling an agent and ending an episode. Theoretically you can disable and re-enable an agent in the same team episode but not with destroy.
Having to set TeamManager for disabled agent every episode (or even within episode) and getting different results from calling GetRegisteredAgents() at different time is what actually bothers me. 

That said, I agree with your concern about consistency and I can change this.